### ISSUE TITLE
Working with 3D

### ISSUE THREAD

**Issue by soumickmj** [anibali/dsntnn#14]
I'm working with 3D volumes, base model is ResNet3D50 (all convolutions are Conv3D) and wants to predict x, y, z co-ordinates. Can you please help with that? Thanks.

**Comment by soumickmj** [comment]
Currently I'm using this with the 3D model:-

outcoords, heatmaps = self.net(vol)
outcoords = outcoords[:,0,:] #to loose extra dim in between
euc_losses = dsntnn.euclidean_losses(outcoords, coordinates) # Per-location euclidean losses
reg_losses = dsntnn.js_reg_losses(heatmaps, coordinates, sigma_t=1.0) # Per-location regularization losses
loss = dsntnn.average_loss(euc_losses + reg_losses) # Combine losses into an overall loss

and commented out line 200 in the dsntnn init (it was complaining as left is 2D tensor, right is 1D)

Is this the correct method to deal with 3D?
Thanks.

**Comment by anibali** [comment]
Instead of commenting out the assert in dsntnn, you should change the shape of the tensors to match each other. So remove the "lose extra dim in between" line, and instead unsqueeze the target so it is of shape B x 1 x 3 (ie `coordinates = coordinates.unsqueeze(1)`). This is the intended way of using the library.

I'm also assuming that somewhere in `self.net` you are using `flat_softmax` and `dsnt`.

**Comment by soumickmj** [comment]
Thanks. I replaced "loose extra dim in between" with unsqueeze and uncommented that assert.
Forward function of the self.net contains the following:-

x = self.model.features(x) #ignoring logits
unnormalized_heatmaps = self.hm_conv(x) # 2. Use a 1x1 conv to get one unnormalized heatmap per location
heatmaps = dsntnn.flat_softmax(unnormalized_heatmaps) # 3. Normalize the heatmaps
coords = dsntnn.dsnt(heatmaps) # 4. Calculate the coordinates
return coords, heatmaps

If something is wrong let me know.
Thanks.

**Comment by anibali** [comment]
That looks pretty reasonable to me.

**Comment by soumickmj** [comment]
Thank you so much.
One problem I'm encountering now is that after some epochs, model parameters are becoming NaN. Which results in NaN in the output.
Is somehow the loss calculation are to blame?
Do you have any suggestions to deal with this issue?
Thanks.

**Comment by anibali** [comment]
That's a fairly generic problem that can be caused by many different things. If you don't know about it already, `torch.autograd.detect_anomaly` could help you find more info (https://pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly). If you're worried that somehow dsntnn is to blame, that's relatively easy to check---assert that there are no NaNs/infs in the output of self.hm_conv, and assert that there are no NaNs/infs in the loss.

**Comment by soumickmj** [comment]
Thank you so much for your help.
In the problem I'm working on, I have a segmentation mask. And the idea is if the predicted coordinate is anywhere within the region (the region which is set to 1 in the mask), its correct (so no loss basically). But when it is outside of the region, then we need to try to pull it in the region. For the same, I was thinking if I send an inverted mask to the average_loss function. What do you think? Or do you have any other better idea maybe?
Sorry for bothering you and I really appreciate and thankful for all your help.
Thanks.

**Comment by anibali** [comment]
I think that for your problem you would probably be better served using a loss of zero for those points rather than excluding them from the loss calculation. A point that's within your segmentation region is "good" and should lower the loss, not be ignored.

Off the top of my head, I would think about implementing it this way:

1. Use `euclidean_losses` and `js_reg_losses` to get two loss tensors (let's call them `elosses` and `jlosses`)
2. Add them together to get a single loss tensor: `losses = elosses + jlosses`
3. Create a mask which is 0 for points that are within the segmentation mask, and 1 elsewhere
4. Apply the mask to zero out losses for the "good" points: `masked_losses = losses * mask`
5. Average the losses to get the combined loss: `loss = losses.mean()`
